oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/authentication/OidcClientRegistrationAuthenticationProvider.java

```diff
@@ -205,7 +205,8 @@ private OidcClientRegistrationAuthenticationToken registerClient(OidcClientRegis
 
 		RegisteredClient registeredClient = this.registeredClientConverter.convert(clientRegistrationAuthentication.getClientRegistration());
 
-		if (StringUtils.hasText(registeredClient.getClientSecret())) {
+		if (StringUtils.hasText(registeredClient.getClientSecret()) &&
+					!ClientAuthenticationMethod.CLIENT_SECRET_JWT.getValue().equals(clientRegistrationAuthentication.getClientRegistration().getTokenEndpointAuthenticationMethod())) {
```

Comment:
We should keep the `client_secret` hashed (depending on configured `PasswordEncoder`) in order to maintain confidentiality. A solution to the fix can be as simple as having the client use the hashed `client_secret` as the key when creating the client assertion. So instead of returning the "clear" `client_secret`, return the hashed `client_secret` in the client registration response. Makes sense?

Reply:
Thank you for your review. Is the comment's content something like this?

![client_secret_jwt_hashed](https://github.com/spring-projects/spring-authorization-server/assets/36355491/95562315-1e8f-4475-8e3f-6d2247f4c7d5)


In the case of client_secret_post and client_secret_basic, it seems to be effective because it uses a plain client_secret at the time of the request, and the authorization server hashes the plain client_secret and compares it with the database.
However, in the case of client_secret_jwt, the hash value is signed and verified, so even if it is hashed, I think the hash value essentially has the same meaning as a plain secret.

If you don't want to store plain secrets in the DB, I think it's better to encrypt them instead of hashing them.
But I don't think it's necessary to go that far.

For example, Keycloak currently stores secrets plainly.
https://github.com/keycloak/keycloak/discussions/8455

Reply:
> Is the comment's content something like this?

Yes. The sequence diagrams exactly depict the flow.